### ISSUE TITLE
flutter doctor newlines have vanished (Windows)

### ISSUE THREAD

**Issue by DanTup** [flutter/flutter#14696]
Pretty sure this looked ok a few days ago!

![Doctor output](https://user-images.githubusercontent.com/1078012/36194921-840deca0-1163-11e8-82a3-eec57d8fae49.png)

**Comment by DanTup** [comment]
Doesn't seem to affect everything, eg. if I run "flutter" the help text is all correctly formatted.

**Comment by DanTup** [comment]
I can't figure this out... The unprintable character shown here is just a newline (`\n`) but it looks like this code hasn't changed in a long time and if I spit out a newline in PoSh by doing ``"`n" | out-host`` it prints correctly.

**Comment by DanTup** [comment]
Same in a normal command window; not isolated to PowerShell.

**Comment by mit-mit** [comment]
I'm seeing the same on my windows machine

**Comment by mit-mit** [comment]
cc @jcollins-g

**Comment by mit-mit** [comment]
I also thought we were replacing the unicode characters on windows?
https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/base/logger.dart#L123

**Comment by DanTup** [comment]
Interesting! I changed that method to replace with docs:

```dart
    stdout.write(message
        .replaceAll('✗', '')
        .replaceAll('✓', '')
    );
```

and still see the problem (but with dots in the message). I'll hack at the message a little and see what I can turn up.

**Comment by DanTup** [comment]
Problem exists even if I replace that method with:

```dart
void writeToStdOut(String message) {
  stdout.write("fake message\n");
}
```

Makes me think that Windows doesn't like `\n` on its own; but it seems like the code around this stuff hasn't changed for a while 😕

**Comment by DanTup** [comment]
This seems like it might be a good opportunity to learn how to use `git bisect`! I can take a look shortly once I've fixed a couple things.

**Comment by DanTup** [comment]
Hmmm, I ran bisect and it ended up blaming commit 58689c796c0d824cfbfc87d850a4dbd26eba2bc0 but I don't see anything related in there. There is an update to `engine.version` but I don't know what that is - does it include anything that could affect this?

**Comment by DanTup** [comment]
Yep, definitely that changeset; it's parent is good.

![Bisect](https://user-images.githubusercontent.com/1078012/36202905-84ac3404-117d-11e8-88c8-883018a5e965.png)

@mit-mit Assuming this change is in the engine (which looks like C++?) I think this is about as far as I can go!

**Comment by mit-mit** [comment]
cc @goderbauer, @chinmaygarde

**Comment by jcollins-g** [comment]
@DanTup it's my understanding on Windows `\n` tends to get translated into the equivalent of `\r\n` by compilers and friends, automatically.  If something in the runtime has changed to where that no longer happens on Windows, it could explain what you're seeing.  You can probably check that's the problem by putting `\r\n` in place of `\n` and seeing whether that "fixes" it.

i don't currently have a windows machine ready to go, if you need help I can spin up a cloud instance or something.

**Comment by DanTup** [comment]
Changing to `\r\n` doesn't seem to have any effect.

Also, check this out - I copy/pasted the first line further down, and the first one output gets the newline rendered and the last one does not!

![Flutter](https://user-images.githubusercontent.com/1078012/36222011-dd4ffc78-11b7-11e8-953f-6147fbed5cce.png)

I can try and some more digging tomorrow (I'm on UK time; finished for today) to see if I can narrow down where the prints suddenly stop working correctly.

Can someone confirm whether the "engine" updated in the changeset I linked above could have any effect on this? Is anything in there being run as part of this (I figured it was just pure Dart, but I tried several times and it seems to only repro after that changeset). I guess I could also try applying some parts of that changeset too.

**Comment by DanTup** [comment]
And just to make things more fun...

![Fun](https://user-images.githubusercontent.com/1078012/36222268-c0c17130-11b8-11e8-8ef9-b9d6b595b7fe.png)

I'll do some more digging tomorrow!

**Comment by jcollins-g** [comment]
[This](https://github.com/flutter/flutter/commit/a4489945436cc42d4aa4787f32e5d2364dfb7f53) might make much of your analysis obsolete -- flutter tools is changing in what SDK is being run and tying it to the engine.  We should retry to see whether this can still be reproduced at head.

**Comment by DanTup** [comment]
Seems even worse on latest:

![Doctor](https://user-images.githubusercontent.com/1078012/36245739-6bc4466e-1224-11e8-9a4e-02ca17bf17e5.png)

The first line is still fine though; so I'll see if I can isolate what's causing it to go from good to bad.

**Comment by DanTup** [comment]
Ok, I got down to this:

![Doctor](https://user-images.githubusercontent.com/1078012/36246343-460a5474-1227-11e8-80e5-2bee7190d157.png)

Reading `engineDartVersion` causes further output to be broken (which on the surface looks kinda crazy). But I bug into that, and it's basically doing this:

```dart
      final File dartSdkBin = engineDirectory.childFile(fs.path.joinAll(dartSdkBinParts));
      final ProcessResult result = processManager.runSync(<String>[dartSdkBin.path, '--version']);
      // https://github.com/dart-lang/sdk/issues/31481
      // We can use the process utils directly when this is fixed instead of parsing stderr.
      _engineDartVersion = result.stderr.trim().replaceAll('Dart VM version: ', '');
```

This goes into `LocalProcessManager.runSync` which has some encoding arguments which aren't supplied so are set to SYSTEM_ENCODING, though as far as I can see they're only used for decoding the std streams (I can't see anything that might have this effect). Eventually this all ends up in something called `_startNative` which starts to get a bit alien to me.

I'll dig a little further to see if I can find out anymore (for ex. is it the process or running an external process that's breaking it, or specifically running the Dart VM) but I don't know how much further I'll understand!

**Comment by DanTup** [comment]
Right, if I change this code:

```dart
final ProcessResult result = processManager.runSync(<String>[dartSdkBin.path, '--version']);
_engineDartVersion = result.stderr.trim().replaceAll('Dart VM version: ', '');
```

to:

```dart
final ProcessResult result = processManager.runSync(<String>['cmd', '/C', 'echo', 'Dart VM version: 2.0.0-edge.3c4dccbd46f152be9e1b6ca95c57357e8e48057c (Wed Feb 14 01:44:18 2018 +0000) on "windows_x64"'], workingDirectory: 'C:\\Windows\\system32');
_engineDartVersion = result.stdout.trim().replaceAll('Dart VM version: ', '');
```

Then the problem goes away. So, I don't think it's the act of spawning the process doing this, but rather the Dart VM doing something that is affecting the stdout stream for the flutter process even after it's finished.

I don't think I can take this any further, but not sure who might be best to look at it (cc @mit-mit @devoncarew).

**Comment by DanTup** [comment]
This might now explain the changeset that `bisect` found - the engine was rolled in that version, and this code is using the engine's version of the SDK. So I'd guess is that there was a change in the SDK that was brought into the engine between f5a4a9378740c3d5996583a9ed1f7e28ff08ee85 and 36a9c343b7c5ec361b60c37d154071146daeeea7 that introduced this.

**Comment by DanTup** [comment]
Here's a simpler repro for testing:

```dart
import "dart:io";

main() async {
  print("Hello\nworld");

  await Process
      .run('M:\\Apps\\Dart\\Dart-2.0.0-dev.24.0\\bin\\dart.exe', ['--version']);

  print("Hello\nworld");

  await Process.run(
      'M:\\Coding\\Applications\\flutter\\bin\\cache\\dart-sdk\\bin\\dart.exe',
      ['--version']);

  print("Hello\nworld");
}
```

The first path is 2.0.0-dev.24.0 and the second path is the current flutter version. Output looks like this:

![Output](https://user-images.githubusercontent.com/1078012/36247903-8c0020f2-122d-11e8-85e4-9ff03968928c.png)

**Comment by DanTup** [comment]
Correction; I misread the results above - it seems to have become broken between Dart-2.0.0-dev.22.0 and Dart-2.0.0-dev.32.0:

![Newline issue](https://user-images.githubusercontent.com/1078012/36250680-71e0347e-1236-11e8-9727-413d648d003f.png)

**Comment by mit-mit** [comment]
Tracking in https://github.com/dart-lang/sdk/issues/32172 now

**Comment by DanTup** [comment]
Btw, is it neceesary to spawn the VM to get the version here? If we have a path to the SDK, there are files inside named `version`, `revision`, etc.?

**Comment by mit-mit** [comment]
Fix landed in Dart; waiting for a roll

**Comment by tvolkert** [comment]
This was fixed by @a-siva in a37a8424b389d82ce42fa369f7a8903fa80e5bcc
